File: bindings/python/aster/interceptors/retry.py
"""Retry interceptor for transient client-side failures."""

from __future__ import annotations

import random

from aster.interceptors.base import CallContext, Interceptor
from aster.status import RpcError, StatusCode
from aster.types import RetryPolicy


class RetryInterceptor(Interceptor):
    """Provides retry policy hints for client calls."""

    def __init__(
        self,
        policy: RetryPolicy | None = None,
        retryable_codes: set[StatusCode] | None = None,
    ) -> None:
        self.policy = policy or RetryPolicy()
        self.retryable_codes = retryable_codes or {StatusCode.UNAVAILABLE}

    def should_retry(self, ctx: CallContext, error: RpcError) -> bool:
        return ctx.idempotent and error.code in self.retryable_codes

    def backoff_seconds(self, attempt: int) -> float:
        backoff = self.policy.backoff
        delay_ms = min(
            backoff.max_ms,
            int(backoff.initial_ms * (backoff.multiplier ** max(0, attempt - 1))),
        )
        jitter = delay_ms * backoff.jitter * random.random()
        return (delay_ms + jitter) / 1000.0

File: bindings/typescript/packages/aster/src/interceptors/retry.ts
/**
 * Retry interceptor with exponential backoff.
 *
 * Only retries idempotent calls on retryable status codes.
 */

import type { Interceptor } from './base.js';
import { CallContext } from './base.js';
import { RpcError, StatusCode } from '../status.js';
import { DEFAULT_RETRY, type RetryPolicy } from '../types.js';

const RETRYABLE_CODES = new Set<number>([
  StatusCode.UNAVAILABLE,
  StatusCode.DEADLINE_EXCEEDED,
  StatusCode.ABORTED,
  StatusCode.RESOURCE_EXHAUSTED,
]);

export class RetryInterceptor implements Interceptor {
  private policy: RetryPolicy;

  constructor(policy?: Partial<RetryPolicy>) {
    this.policy = { ...DEFAULT_RETRY, ...policy };
  }

  async onError(ctx: CallContext, error: RpcError): Promise<RpcError> {
    if (!ctx.idempotent) return error;
    if (!RETRYABLE_CODES.has(error.code)) return error;
    if (ctx.attempt >= this.policy.maxAttempts) return error;

    // Signal retry by attaching metadata
    error.details['retry_after_ms'] = String(this.backoffMs(ctx.attempt));
    error.details['retry_attempt'] = String(ctx.attempt);
    return error;
  }

  private backoffMs(attempt: number): number {
    const { initialMs, multiplier, maxMs, jitter } = this.policy.backoff;
    const base = Math.min(initialMs * Math.pow(multiplier, attempt - 1), maxMs);
    const jitterAmount = base * jitter * (Math.random() * 2 - 1);
    return Math.max(0, Math.round(base + jitterAmount));
  }

  /**
   * Decide if a call should be retried.
   * Returns true if the error is retryable, call is idempotent, and attempts remain.
   */
  shouldRetry(error: RpcError, attempt: number, idempotent = false): boolean {
    if (!idempotent) return false;
    if (!RETRYABLE_CODES.has(error.code)) return false;
    return attempt < this.policy.maxAttempts;
  }

  /**
   * Compute backoff duration in seconds for the given attempt number.
   */
  backoffSeconds(attempt: number): number {
    return this.backoffMs(attempt) / 1000;
  }
}
